### PR TITLE
VMSS: loadbalancer rules, public ip address, NAT rules.

### PR DESCRIPTION
**Pull request by sergey-shandar**
## Description

VMSS: loadbalancer rules, public ip address, NAT rules.

## Checklist

- [x] I have read the [_Submitting Changes_](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#making-changes) section of [`CONTRIBUTING.md`](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate [change log has been updated](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log)
- [ ] The PR does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] the changes have gone through a [cmdlet design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr)
    - [ ] the cmdlet markdown files were [generated using the `platyPS` module](https://github.com/Azure/azure-powershell/blob/preview/documentation/development-docs/help-generation.md)

### CODE REVIEW

File: src/ResourceManager/Compute/Commands.Compute/Manual/VirtualMachineScaleSetCreateOrUpdateMethod.cs [hunk 1]
```diff
@@ -114,6 +107,9 @@ public partial class NewAzureRmVmss : ComputeAutomationBaseCmdlet
             ValueFromPipelineByPropertyName = true)]
         public List<string> Zone { get; set; }
 
+        [Parameter(ParameterSetName = SimpleParameterSet, Mandatory = false)]
+        public int[] NatBackendPort { get; set; } = null;
```
Comment (markcowl) [inline src/ResourceManager/Compute/Commands.Compute/Manual/VirtualMachineScaleSetCreateOrUpdateMethod.cs]: don't think we need to initialize to null

File: src/ResourceManager/Compute/Commands.Compute/Strategies/ComputeRp/SubResourceStrategy.cs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+{
+    static class SubResourceStrategy
+    {
+        static SubResource GetSuResourceReference(
+            this IEngine engine, IEntityConfig config)
+            => config == null ? null : new SubResource { Id = engine.GetId(config) };
+
+        public static SubResource GetReference(
+            this IEngine engine, ResourceConfig<AvailabilitySet> availabilitySet)
+            => engine.GetSuResourceReference(availabilitySet);
```
Comment (markcowl) [inline src/ResourceManager/Compute/Commands.Compute/Strategies/ComputeRp/SubResourceStrategy.cs]: should this method name be GetSubResourceReference?

File: src/ResourceManager/Compute/Commands.Compute/Manual/VirtualMachineScaleSetCreateOrUpdateMethod.cs [hunk 3]
```diff
@@ -174,6 +183,21 @@ public partial class NewAzureRmVmss : ComputeAutomationBaseCmdlet
...
+            var portRangeStart = 50000;
+            foreach (var natBackendPort in NatBackendPort)
+            {                
+                var inboundNutRule = loadBalancer.CreateInboundNatPool(
+                    name: inboundNutRuleName + natBackendPort.ToString(),
+                    frontendIpConfiguration: frontendIpConfiguration,
+                    frontendPortRangeStart: portRangeStart,
+                    frontendPortRangeEnd: portRangeStart + InstanceCount,
+                    backendPort: natBackendPort);
+                portRangeStart += 1000;
```
Comment (markcowl) [inline src/ResourceManager/Compute/Commands.Compute/Manual/VirtualMachineScaleSetCreateOrUpdateMethod.cs]: How many rules could we have?  I wonder if we should just fail gracefully if #instances > 1000

File: src/ResourceManager/Compute/Commands.Compute/Strategies/ComputeRp/SubResourceStrategy.cs (hunk 2 above)
Comment (sergey-shandar) [reply src/ResourceManager/Compute/Commands.Compute/Strategies/ComputeRp/SubResourceStrategy.cs]: The idea is to use always the same function name for all resource to get a reference. Different overloads will return different results. For example:
- Network: 
  - uses the same types for a resource Id wrapper object. So the function is `TModel GetReference(ResourceConfig<TModel>)`
  - uses `SubResource` for a nested resource Id wrapper object. `TModel GetReference(NestedResourceConfig<TModel>) where TModel : SubResource`
- Compute uses all kind of Id wrapper objects for different resource types, so I provided several `GetReference` overloads.

File: src/ResourceManager/Compute/Commands.Compute/Manual/VirtualMachineScaleSetCreateOrUpdateMethod.cs (hunk 3 above)
Comment (sergey-shandar) [reply src/ResourceManager/Compute/Commands.Compute/Manual/VirtualMachineScaleSetCreateOrUpdateMethod.cs]: #5657

**Comment by markcowl** [comment]
on demand run here: https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/powershell-demand/546/

**Pull request merged**